src/ecmult_gen_impl.h

```diff
@@ -200,8 +200,15 @@ static void secp256k1_ecmult_gen(const secp256k1_ecmult_gen_context *ctx, secp25
              * ((block*COMB_TEETH + tooth)*COMB_SPACING + comb_off) of d. */
             uint32_t bits = 0, sign, abs, index, tooth;
             for (tooth = 0; tooth < COMB_TEETH; ++tooth) {
-                uint32_t bit = (recoded[bit_pos >> 5] >> (bit_pos & 0x1f)) & 1;
-                bits |= bit << tooth;
+                /* Instead of reading individual bits here to construct bits, build up
+                 * the result by xoring shifted reads together. In every iteration, one
+                 * additional bit is made correct, starting at the bottom. The bits
+                 * above that contain junk. This reduces leakage from single bits. See
+                 * https://www.usenix.org/system/files/conference/usenixsecurity18/sec18-alam.pdf
+                 */
+                uint32_t bitdata = recoded[bit_pos >> 5] >> (bit_pos & 0x1f);
+                bits &= ~(1 << tooth);
+                bits ^= bitdata << tooth;
```

Comment:
Hm, this is difficult to analyze and review. 

If `(bit_pos & 0x1f)` happens to be 31, we still extract a single bit. Could it help to ensure that there's always some junk in the high bits of `bitdata`? (Not sure what's the best way to achieve this.)

--
Also, compilers are not helpful here: gcc keeps the xor, but clang 17.0.6 manages to move the xor out of the loop and replace by a mov.

To inspect this, try `objdump-nice .libs/libsecp256k1.so | less` and search for `bitdata`, where `objdump-nice` is a shell alias for `objdump --disassemble --wide --debugging --line-numbers --source-comment --visualize-jumps=extended-color --disassembler-options=intel --no-show-raw-insn --no-addresses`

`volatile` helps, but I'm not convinced:
 - `bits = *((volatile uint32_t*)&bits) ^ (bitdata << tooth);` keeps the xor, but moves `bits` from a register to memory. Not sure if this better in terms of leakage.
 - `bits ^= bitdata << *((volatile uint32_t*)&tooth);` also keeps the xor. This moves `tooth` to memory and prevents clang from unrolling the loop. 

--

By the way, the mitigation in OpenSSL is https://github.com/openssl/openssl/pull/6276, but looking at that commit doesn't seem particularly helpful in our case.

Reply:
See https://github.com/real-or-random/secp256k1/commit/50ef2da3df16e7e610098bd50cacb60e5264cf7c for an improved version. I've manually verified that gcc, clang, msvc actually use rots and don't optimize that away.